middle_end/flambda2/types/grammar/type_grammar.ml

```diff
@@ -132,14 +132,16 @@ and closures_entry =
 
 (* Products are a set of constraints: each new field reduces the concrete set.
    The empty product is top. There is no bottom. All components must be of the
-   same kind.
+   same kind except for [value_slot_indexed_product].
 
    { 1 => Unknown; 2 => V } is equal to { 2 => V } *)
 and function_slot_indexed_product =
   { function_slot_components_by_index : t Function_slot.Map.t }
 
 and value_slot_indexed_product =
-  { value_slot_components_by_index : t Value_slot.Map.t }
+  { value_slot_components_by_index :
+      (t * Flambda_kind.With_subkind.t) Value_slot.Map.t
+  }
```

Comment:
It's not clear to me why it is useful to add the subkind here. The type already contains a kind.

Reply:
From Slack (from @lthls ): ... the layout computation only looks at the set of closure creations, so as long as we have full kinds and subkinds there we should be fine. If we started reifying closures it could be useful, but I'm not convinced it would be necessary even in this case